lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -0,0 +1,235 @@
+import { StaticJsonRpcProvider, TransactionRequest } from '@ethersproject/providers'
+import { Config, DEFAULT_CONFIG } from './config'
+import { ID_TO_NETWORK_NAME, metric, MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { ChainId } from '@uniswap/sdk-core'
+import {
+  BlockTag,
+  BlockWithTransactions,
+  Filter,
+  Log,
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/abstract-provider'
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Deferrable } from '@ethersproject/properties'
+import { deriveProviderName } from '../handlers/evm/provider/ProviderName'
+import Logger from 'bunyan'
+
+interface SingleCallPerf {
+  succeed: boolean
+  latencyInMs: number
+  startTimestampInMs: number
+}
+
+// TODO(jie): Implement block-aligned cache
+export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
+  readonly url: string
+  readonly providerName: string
+
+  private healthScore
+  private healthy: boolean
+  private lastCallTimestampInMs: number
+  private config: Config
+  private readonly metricPrefix: string
+  private readonly log: Logger
+
+  constructor(chainId: ChainId, url: string, log: Logger, config: Config = DEFAULT_CONFIG) {
+    super(url, { chainId, name: ID_TO_NETWORK_NAME(chainId) })
```

Comment:
nit: the existing `network?: Networkish` within `InstrumentedEVMProviderProps` is more strongly-typed than the anonymous typed object `{ chainId, name: ID_TO_NETWORK_NAME(chainId) }` here. I would write as:

```
constructor(network: Networkish, url: string, log: Logger, config: Config = DEFAULT_CONFIG) {
    super(url, network)
    this.url = url
    this.log = log
    this.providerName = deriveProviderName(url)
    this.healthScore = 0
    this.healthy = true
    this.lastCallTimestampInMs = 0
    this.config = config
    this.metricPrefix = `RPC_${this.providerName}_${this.network.chainId}`
  }
```

meanwhile you don't need to change the caller code, because `Networkish` has a union type of number.